impl/otbn/rsa/mont_rr.i.vad

```diff
@@ -0,0 +1,318 @@
+include "../../../spec/arch/otbn/decls.i.vad"
+include "montmul.i.vad"
+
+#verbatim
+include "../../../spec/arch/otbn/decls.i.dfy"
+include "../../../../glue/otbn/rsa/bv256_mm_lemmas.i.dfy"
+
+include "cond_sub_to_reg.i.dfy"
+include "montmul.i.dfy"
+
+module mont_rr {
+
+import opened ot_machine
+import opened ot_abstraction
+import opened ot_decls
+import opened ot_vale
+
+import opened cond_sub_to_reg
+import opened montmul
+
+import opened integers
+import opened bv256_mm_lemmas
+
+#endverbatim
+
+function xor_clear_lemma(x: uint256) : void extern;
+
+function xor_negation_lemma(x: uint32) : void extern;
+
+function addc_inv(dst: seq(uint256),
+    carry: uint1,
+    src1: seq(uint256),
+    src2: seq(uint256),
+    index: int) : bool extern;
+
+ghost procedure addc_inv_pre_lemma(
+    ghost dst: seq(uint256),
+    ghost src1: seq(uint256),
+    ghost src2: seq(uint256)) extern;
+
+ghost procedure addc_inv_peri_lemma(
+    ghost dst: seq(uint256),
+    ghost new_carry: uint1,
+    ghost src1: seq(uint256),
+    ghost src2: seq(uint256),
+    ghost old_carry: uint1,
+    ghost index: int) extern;
+
+ghost procedure addc_inv_post_lemma(
+    ghost dst: seq(uint256),
+    ghost carry: uint1,
+    ghost src1: seq(uint256),
+    ghost src2: seq(uint256)) extern;
+
+function double_modulo_pre(
+    a: seq(uint256),
+    rsa: rsa_params) : bool extern;
+
+function double_modulo_post(
+    a: seq(uint256),
+    aa: seq(uint256),
+    rsa: rsa_params) : bool extern;
+
+function double_modulo_select_nosub_lemma(
+    a: nat,
+    aa: nat,
+    carry: uint1,
+    borrow: uint1,
+    select: uint32,
+    M: nat) : void extern;
+
+function double_modulo_select_withsub_lemma(
+    a: nat,
+    aa: nat,
+    carry: uint1,
+    borrow: uint1,
+    select: uint32,
+    M: nat) : void extern;
+
+procedure double_mod_copy_result(
+    ghost src_view: seq(uint256))
+    returns (ghost dst_view: seq(uint256))
+
+    requires
+        seq_len(src_view) == NUM_WORDS;
+        valid_wdr_view(wdrs, src_view, 16, NUM_WORDS);
+
+    reads
+        w16; w17; w18; w19; w20; w21; w22; w23; w24; w25; w26; w27;
+
+        // This does not actually read w31, but for some reason I can't prove
```

Comment:
Good observation. This is actually due to `wdr` in the `modifies` clause. `wdr` refers to all the wide registers at once.Then the caller is not clear on which registers are updated specifically, so it can't assume `w31` is unchanged.